clang/lib/Frontend/IncludeTreePPActions.cpp

```diff
@@ -116,65 +116,85 @@ class IncludeTreePPActions final : public PPCachedActions {
+        -> std::variant<std::monostate, IncludeFile> {
+      auto File = EnteredTree.getBaseFile();
+      if (!File)
+        return reportError(File.takeError());
+      auto FilenameBlob = File->getFilename();
+      if (!FilenameBlob)
+        return reportError(FilenameBlob.takeError());
+
+      SourceManager &SM = PP.getSourceManager();
+      Expected<FileEntryRef> FE =
+          SM.getFileManager().getFileRef(FilenameBlob->getData(),
+                                         /*OpenFile=*/true);
+      if (!FE)
+        return reportError(FE.takeError());
+      FileID FID =
+          SM.createFileID(*FE, IncludeLoc, EnteredTree.getFileCharacteristic());
+      PP.markIncluded(*FE);
+      IncludeStack.push_back(
+          {std::move(EnteredTree), SM.getLocForStartOfFile(FID)});
+
+      Module *M = nullptr;
+      auto SubmoduleName = EnteredTree.getSubmoduleName();
+      if (!SubmoduleName)
+        return reportError(SubmoduleName.takeError());
+      if (*SubmoduleName) {
+        SmallVector<StringRef> ModuleComponents;
+        (*SubmoduleName)->split(ModuleComponents, '.');
+        M = PP.getHeaderSearchInfo().lookupModule(
+            ModuleComponents[0], IncludeLoc,
+            /*AllowSearch=*/false, /*AllowExtraModuleMapSearch=*/false);
         if (!M)
-          return reportErrorTwine(
-              llvm::Twine("failed to find or infer submodule '") + Sub + "'");
+          return reportErrorTwine(llvm::Twine("failed to find module '") +
+                                  ModuleComponents[0] + "'");
+        for (StringRef Sub : ArrayRef(ModuleComponents).drop_front()) {
+          M = M->findOrInferSubmodule(Sub);
+          if (!M)
+            return reportErrorTwine(
+                llvm::Twine("failed to find or infer submodule '") + Sub + "'");
+        }
+
+        // Add to known headers for the module.
+        ModuleMap &MMap = PP.getHeaderSearchInfo().getModuleMap();
+        Module::Header H{"", "", *FE};
+        MMap.addHeader(M, std::move(H), ModuleMap::NormalHeader);
       }
 
-      // Add to known headers for the module.
-      ModuleMap &MMap = PP.getHeaderSearchInfo().getModuleMap();
-      Module::Header H{"", "", *FE};
-      MMap.addHeader(M, std::move(H), ModuleMap::NormalHeader);
-    }
+      return IncludeFile{FID, M};
+    };
 
-    return IncludeFile{FID, M};
+    switch (Node->getKind()) {
+    case cas::IncludeTree::NodeKind::ModuleImport:
+      return MakeModuleImport(Node->getModuleImport());
+    case cas::IncludeTree::NodeKind::Tree: {
+      auto IncludeTree = MakeIncludeTree(Node->getIncludeTree());
+      assert(std::holds_alternative<IncludeFile>(IncludeTree));
+      return std::get<IncludeFile>(IncludeTree);
+    }
+    case cas::IncludeTree::NodeKind::SpuriousImport:
+      auto SpuriousImportNode = Node->getSpuriousImport();
+      auto ModuleImport = MakeModuleImport(
+          llvm::cantFail(SpuriousImportNode.getModuleImport()));
```

Comment:
Why `cantFail` (same below)?  If there's a CAS error this will assert or be UB in a release build